### PR TITLE
Non critical triggered schedulers

### PR DESCRIPTION
**Pull request by Riziero**
capture

### CODE REVIEW

**Comment by Riziero** [comment]
in the shot you can see that an unimportant scheduler wont cause a fail of the triggering step.

**Comment by Riziero** [comment]
anybody know what's wrong with the docs? Why's it failing?

File: master/buildbot/steps/trigger.py [hunk 1]
```diff
@@ -198,13 +210,21 @@ def addBuildUrls(self, rclist):
     @defer.inlineCallbacks
     def run(self):
         schedulers_and_props = yield self.getSchedulersAndProperties()
+        
+        # To be back compatible we need to differ between old and new style
+        # schedulers_and_props can either consist of 2 elements or 3 elements tuples
+        
+        # If the user getSchedulersAndProperties funcion does not specify the importance (red len(tup)==2), we make all schedulers important
+        schedulers_and_props = [tup if len(tup)==3 else (tup[0], tup[1], False) for tup in schedulers_and_props]
```
Comment (tardyp) [inline master/buildbot/steps/trigger.py]: I think it would be better to change to a dict or better, a namedtuple so that futureproof is even easier.
https://docs.python.org/3/library/collections.html#collections.namedtuple

If we need to add another argument we would need if len == 4, which will be even worse.
I usually dont like too much if-in-the-future arguments, but in this case this choice wont be undoable as this is API.

Another option is to add self.getSchedulersAndPropertiesWithImportance, which would call the older version, but this is not very clean either.

Comment (Riziero) [inline master/buildbot/steps/trigger.py]: so you are suggesting something like:

if tuple:
  old_style
else:
 #dict
?

**Comment by sa2ajj** [comment]
re failing docs: you use a new name that is not part of the dictionary. you probably want to check formatting of your docs update.

File: master/buildbot/steps/trigger.py (hunk 1 above)
Comment (tardyp) [inline master/buildbot/steps/trigger.py]: I would rather say if dict: break else: convert_old_style

File: master/buildbot/steps/trigger.py [hunk 2]
```diff
@@ -199,12 +225,29 @@ def addBuildUrls(self, rclist):
     def run(self):
         schedulers_and_props = yield self.getSchedulersAndProperties()
 
+        schedulers_and_props_list = []
+        
+        # To be back compatible we need to differ between old and new style
+        # schedulers_and_props can either consist of 2 elements tuple or dictionary
+        for element in schedulers_and_props:
+            if type(element) == dict:
+                schedulers_and_props_list.append(element)
```
Comment (tardyp) [inline master/buildbot/steps/trigger.py]: I would do short path:
schedulers_and_props_list = schedulers_and_props
break

**Comment by Riziero** [comment]
guys I've got no idea how to fix the docs. Could you give a lead? Thanks :)

**Comment by tardyp** [comment]
You can use "make docschecks" to run locally those tests

Then you can look at the other doc PR to see how to change the spelling file
https://github.com/buildbot/buildbot/pull/2300/files

**Comment by Riziero** [comment]
Getting this:

Running Sphinx v1.4.3
Exception occurred:
  File "conf.py", line 35, in <module>
    raise RuntimeError("sphinxcontrib.blockdiag is not installed. "
RuntimeError: sphinxcontrib.blockdiag is not installed. Please install documentation dependencies with `pip install buildbot[docs]`
The full traceback has been saved in /tmp/sphinx-err-ecxuiN.log, if you want to report the issue to the developers.

pip install buildbot[docs] says I alrady have everything I need.

**Comment by tardyp** [comment]
@Riziero you have to upgrade pip in order to not have the recursive dep bug

pip install -U pip

**Comment by Riziero** [comment]
@tardyp Now getting:

loading pickled environment... not yet created

Sphinx error:
Builder name spelling not registered
make[1]: *** [spelling] Error 1
make[1]: Leaving directory `/home/vagrant/work/master/docs'
make: *** [docschecks] Error 2

Sorry to be a bummer

**Comment by tardyp** [comment]
https://bitbucket.org/dhellmann/sphinxcontrib-spelling/issues/6/builder-name-spelling-not-registered

Is it possible there's a missing C library needed by PyEnchant? If there's a wheel-based package we wouldn't see a failure at installation time but would at runtime when the import fails. Getting the traceback from sphinx would be useful. If that's hard, you could add something like "python -c 'import sphinxcontrib.spelling'" as a command in tox.ini before running sphinx properly.

**Comment by Riziero** [comment]
@tardyp  sudo apt-get install libenchant1c2a did the trick

File: master/buildbot/steps/trigger.py [hunk 3]
```diff
@@ -44,12 +44,14 @@ class Trigger(BuildStep):
 
     flunkOnFailure = True
 
-    def __init__(self, schedulerNames=None, sourceStamp=None, sourceStamps=None,
+    def __init__(self, schedulerNames=None, unimportantSchedulerNames=None, sourceStamp=None, sourceStamps=None,
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: my suggestion would be to add this new parameter before `**kwargs`: while all over the place we promote explicit parameter naming, some users might "rely" on a particular order of parameters.

File: master/buildbot/steps/trigger.py [hunk 4]
```diff
@@ -63,7 +65,17 @@ def __init__(self, schedulerNames=None, sourceStamp=None, sourceStamps=None,
             config.error(
                 "You can't specify both alwaysUseLatest and updateSourceStamp"
             )
+        if len(set(schedulerNames) & set(unimportantSchedulerNames)) != len(unimportantSchedulerNames):
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: there's an easier way to check this:
```python
if not set(schedulerNames).issuperset(set(unimportantSchedulerNames)):
 ...
```

(I'm using `superset` here to make it easier to read the code)

File: master/buildbot/steps/trigger.py [hunk 5]
```diff
@@ -129,7 +141,17 @@ def getSchedulerByName(self, name):
     # This customization enpoint allows users to dynamically select which
     # scheduler and properties to trigger
     def getSchedulersAndProperties(self):
-        return [(sched, self.set_properties) for sched in self.schedulerNames]
+
+            schedulers_and_props = []
+            for sched in self.schedulerNames:
+                d = {
+                        'sched_name': sched,
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: please de-dent these three lines

File: master/buildbot/steps/trigger.py [hunk 6]
```diff
@@ -129,7 +141,17 @@ def getSchedulerByName(self, name):
     # This customization enpoint allows users to dynamically select which
     # scheduler and properties to trigger
     def getSchedulersAndProperties(self):
-        return [(sched, self.set_properties) for sched in self.schedulerNames]
+
+            schedulers_and_props = []
+            for sched in self.schedulerNames:
+                d = {
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: is there a reason to introduce this temporary variable?  its introduction does not seem to help anything.

File: master/buildbot/steps/trigger.py [hunk 7]
```diff
@@ -63,7 +65,17 @@ def __init__(self, schedulerNames=None, sourceStamp=None, sourceStamps=None,
...
+            config.error(
+                "unimportantSchedulerNames must be a subset of schedulerNames"
+            )
+        if len(schedulerNames) <= len(unimportantSchedulerNames):
+            config.error(
+                    "SchedulerNames must at least contain an important scheduler"
+            )
+
         self.schedulerNames = schedulerNames
+        self.unimportantSchedulerNames = unimportantSchedulerNames
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: should we store a set instead? (it's already used above and then it'd be handy in the code below)

File: master/buildbot/steps/trigger.py [hunk 8]
```diff
@@ -167,14 +189,18 @@ def getAllGotRevisions(self):
...
             if isinstance(results, tuple):
-                results, _ = results
+                results, brids_dict = results
 
             if not was_cb:
                 yield self.addLogWithFailure(results)
                 results = EXCEPTION
+
+            # continue if this brid has to be ignored
+            if len(set(unimportant_brids) & set(brids_dict.values())) != 0:
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: do you ignore the whole set if *any* of them were unimportant?

File: master/buildbot/steps/trigger.py [hunk 9]
```diff
@@ -199,12 +225,30 @@ def addBuildUrls(self, rclist):
     def run(self):
         schedulers_and_props = yield self.getSchedulersAndProperties()
 
+        schedulers_and_props_list = []
+
+        # To be back compatible we need to differ between old and new style
+        # schedulers_and_props can either consist of 2 elements tuple or dictionary
+        for element in schedulers_and_props:
+            if type(element) == dict:
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: `isinstance(element, dict)`

File: master/buildbot/steps/trigger.py [hunk 10]
```diff
@@ -199,12 +225,30 @@ def addBuildUrls(self, rclist):
...
+
         # post process the schedulernames, and raw properties
         # we do this out of the loop, as this can result in errors
         schedulers_and_props = [(
-            self.getSchedulerByName(sch),
-            self.createTriggerProperties(props_to_set))
-            for sch, props_to_set in schedulers_and_props]
+            self.getSchedulerByName(entry_dict['sched_name']),
+            self.createTriggerProperties(entry_dict['props_to_set']),
+            entry_dict['unimportant'])
```
Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: wrong indentation

File: master/buildbot/steps/trigger.py (hunk 3 above)
Comment (tardyp) [inline master/buildbot/steps/trigger.py]: +1 good point

File: master/buildbot/steps/trigger.py (hunk 8 above)
Comment (tardyp) [inline master/buildbot/steps/trigger.py]: It is indeed a complexity of the difference between buildsets and buildrequests.

In this particular case schedulers are marked unimportants. There is one bsid by scheduler, and then theorically one brid by builder (in the case the scheduler triggers several builders).
For the usecase we are working on this is highly unlikely, but we should take care of this corner case though.

So I think this line is correct (it should use superset though), but it might deserve some explaination in comment

File: master/buildbot/test/unit/test_steps_trigger.py [hunk 11]
```diff
@@ -542,3 +550,23 @@ def test_waitForFinish_interrupt(self):
...
+        self.expectTriggeredWith(a=(False, [], {}), b=(False, [], {}))
+        return self.runStep()
+
+    def test_unimportantsShedulerNames(self):
+        self.setupStep(trigger.Trigger(schedulerNames=['a', 'b'], unimportantSchedulerNames=['b']))
+        self.scheduler_a.result = SUCCESS
+        self.scheduler_b.result = FAILURE
+        self.expectOutcome(result=SUCCESS, state_string='triggered a, b')
+        self.expectTriggeredWith(a=(False, [], {}), b=(False, [], {}))
+        return self.runStep()
```
Comment (tardyp) [inline master/buildbot/test/unit/test_steps_trigger.py]: We should also test the case where one of the scheduler runs several builders

File: master/docs/manual/cfg-buildsteps.rst [hunk 12]
```diff
@@ -2443,6 +2443,14 @@ Hyperlinks are added to the build detail web pages for each triggered build.
 
         It is possible, but not advisable, to create a cycle where a build continually triggers itself, because the schedulers are specified by name.
 
+``unimportantSchedulerNames``
+    When ``waitForFinish`` is ``True``, all schedulers in this list will not cause the trigger step to fail. unimportantSchedulerNames must be a subset of schedulerNames
+    If ``waitForFinish`` is ``False``, unimportantSchedulerNames will simply be ignored.
+    
+.. note::
+
+        ``schedulerNames`` must have at least one scheduler not present in ``unimportantSchedulerNames``
```
Comment (tardyp) [inline master/docs/manual/cfg-buildsteps.rst]: I am not sure it is a requirement.

If all schedulers are unimportants, then the result of the trigger step is always success, provided that the schedulers exist

**Comment by tardyp** [comment]
thanks @sa2ajj for the great review and @Riziero for the great improvements. A little bit effort remaining and we can land this!

File: master/buildbot/steps/trigger.py (hunk 8 above)
Comment (Riziero) [inline master/buildbot/steps/trigger.py]: what I had in mind: if a scheduler is marked as unimportant, then all the builders this scheduler might kick do not affect the result os the step.

File: master/docs/manual/cfg-buildsteps.rst (hunk 12 above)
Comment (Riziero) [inline master/docs/manual/cfg-buildsteps.rst]: @tardyp If all schedulers are unimportants. This can not happen. There must at least be an important schduler.

File: master/buildbot/steps/trigger.py (hunk 10 above)
Comment (Riziero) [inline master/buildbot/steps/trigger.py]: @sa2ajj can you elaborate on this? it's how it was before in terms of indentation.

Comment (sa2ajj) [inline master/buildbot/steps/trigger.py]: Sorry, my mistake: I misread it as a sequence of method calls, while it's a list comprehension.

File: master/buildbot/test/unit/test_steps_trigger.py (hunk 11 above)
Comment (Riziero) [inline master/buildbot/test/unit/test_steps_trigger.py]: @tardyp setupStep how would you about this? SetupStep only setups 2 scheds with 1 build/buildrequest each. Should I override it?

File: master/docs/manual/cfg-buildsteps.rst (hunk 12 above)
Comment (Riziero) [inline master/docs/manual/cfg-buildsteps.rst]: Bump @tardyp

Comment (tardyp) [inline master/docs/manual/cfg-buildsteps.rst]: Why do you say there must be at least one important scheduler.

I can see usecase where I want to trigger lots of parallel tests just for information, and not take the result in account for the global step status

Comment (Riziero) [inline master/docs/manual/cfg-buildsteps.rst]: you're right. After switching to super set SchedulerNames and unimportantSchedulerNames can actually match.

File: master/buildbot/test/unit/test_steps_trigger.py (hunk 11 above)
Comment (Riziero) [inline master/buildbot/test/unit/test_steps_trigger.py]: bump ;)

File: master/docs/manual/cfg-buildsteps.rst (hunk 12 above)
Comment (Riziero) [inline master/docs/manual/cfg-buildsteps.rst]: that also means that the following code (and its test) must go

 if len(schedulerNames) <= len(unimportantSchedulerNames):
         config.error(
                 "SchedulerNames must at least contain an important scheduler"
         )

File: master/buildbot/test/unit/test_steps_trigger.py (hunk 11 above)
Comment (Riziero) [inline master/buildbot/test/unit/test_steps_trigger.py]: @tardyp how would you go about this? SetupStep only setups 2 scheds with 1 build/buildrequest each. Should I override it?

File: master/docs/manual/cfg-buildsteps.rst (hunk 12 above)
Comment (tardyp) [inline master/docs/manual/cfg-buildsteps.rst]: indeed

File: master/buildbot/test/unit/test_steps_trigger.py (hunk 11 above)
Comment (tardyp) [inline master/buildbot/test/unit/test_steps_trigger.py]: You can just create another scheduler 'c' in SetupStep which only one unit test will use

**Comment by Riziero** [comment]
can we land this or some more stuff to polish? Thanks!
@tardyp

**Comment by tardyp** [comment]
Please clean the git history. Git rebate --help

**Comment by Riziero** [comment]
what do you mean by clening up the history? Squashing all commits into one? Pardon my git knwoledge is very rudimental.

**Comment by tardyp** [comment]
Yes please

**Comment by Riziero** [comment]
@tardyp been trying for 2 hours with no luck - I ended up messing up my branch with 2 more stupid commits, one of them being just empty.

**Comment by tardyp** [comment]
Sorry, I was afk here is the manual
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

This is important for github contributor to know this very basic git tricks. I would not do you a favor to do it for you.

**Comment by Riziero** [comment]
Hey @tardyp 

I am really trying my best. All I am doing from cmd after switching to my WIP branch is

git rebase -i -k e925c49e4d752ce82b8b1b4defe552dc8f856e8 with this hash being last commit not made by me. Basically if I understand things rightly I am telling git to collapse all the following commits into a signle one. 

image

After I save and quit, I get an error about which I found http://git.661346.n2.nabble.com/Confusing-error-message-in-rebase-when-commit-becomes-empty-td7612948.html

**Comment by tardyp** [comment]
You can hang on irc for live help.

basically you should do this in case of empty commits:
git rebase --skip

**Comment by tardyp** [comment]
good work! thanks for keeping up!

**Comment by Riziero** [comment]
@tardyp I was expecting this. When I forked nineRC1 release notes hadnt yet been made. As per guidelines I added my rel note line to index.rst which has changed a lot. hence the conflict.

How do I go about this?
Thanks

**Comment by tardyp** [comment]
looks good! thanks!
